### ISSUE TITLE
Retrying requests at http error 400

### ISSUE THREAD

**Issue by pkochubey** [tus/tusd#317]
**Question**
In documentation you write: 

> Tusd uses the Pester library to issue requests and handle retries. By default, tusd will retry 3 times on a 500 Internal Server Error response or network error

It's ok. But what happen if at 'post-finish' you received http Error 400, "game over"? I use v0.13.3 and I see that at error 400 at post-finish, uploading will be dropped, but at 500 I see retrying.

**Comment by Acconut** [comment]
A 4XX response indicates that the request is semantically invalid or malformed. Retrying such a request would not fix the in-validness but just lead to the same 4XX response. Therefore, there is no benefit in retrying except for wasting time.

A 5XX response (e.g. failed to establish database connection) are caused by the server on its own and can be fixed by retrying. The retries make sense here.

Not every HTTP server follows these rules strictly but in most cases those should apply. So yes, there will be no retries on 4XX responses.

**Comment by pkochubey** [comment]
> Not every HTTP server follows these rules strictly but in most cases those should apply. 

Yep this right, sometimes nginx dropped connection with http error 400, in error log only one row: "client prematurely closed stream: only 0 out of 318 bytes of request body received"

**Comment by pkochubey** [comment]
Do you have any idea why nginx write about this issue and point to client?: "client prematurely closed stream: only 0 out of 318 bytes of request body received"

**Comment by Acconut** [comment]
> Yep this right, sometimes nginx dropped connection with http error 400, in error log only one row: "client prematurely closed stream: only 0 out of 318 bytes of request body received"

That may be true, I have never come across this problem. However, since errors such as 404 or 403 should not be retried, I to not think it is a good to idea to retry 4XX by default. Maybe you can modify Nginx to return different status codes in these situations?

**Comment by pkochubey** [comment]
> Maybe you can modify Nginx to return different status codes in these situations?

Yes, I guess this is should be right solution, I will be investigate it.

**Comment by gajus** [comment]
> A 4XX response indicates that the request is semantically invalid or malformed. Retrying such a request would not fix the in-validness but just lead to the same 4XX response. Therefore, there is no benefit in retrying except for wasting time.

Contrary to what you/ documentation says, post-finish returning 400 is being retried:

```
[tusd] 2020/05/10 04:44:41 event="HookInvocationError" type="post-finish" id="0b3774b83eb7b140b297fcf38d10fd12" error="endpoint returned: 400 Bad Request"
[tusd] 2020/05/10 04:44:42 event="HookInvocationError" type="post-finish" id="9c1afb23c713443350d75835f03ff8bb" error="endpoint returned: 400 Bad Request"
[tusd] 2020/05/10 04:44:42 event="HookInvocationError" type="post-finish" id="a20db6e97800b32c2f072cc837d88782" error="endpoint returned: 400 Bad Request"

```

**Comment by Acconut** [comment]
@gajus That's unfortunate. Could you open an issue to fix the documentation or alternative improve the retry handling?
